 # ENTREGA TABAJO FINAL CODER HOUSE
 * Defunciones viales (archivo.csv)
 * Objetivo de la entrega porder comparar diferentes modelos de prediccion
 * variable objetivo: Clase victima
 * comparar diferentes tipos d modelos.

In [5]:
import pandas as pd

# Cargar el archivo CSV con punto y coma como delimitador y codificación 'latin1'
df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='latin1')


In [6]:
# EDA
# Ver las primeras filas del dataframe
print(df.head())

# Descripción estadística
print(df.describe())

# Información sobre tipos de datos y valores nulos
print(df.info())


                    fuente numero victima  municipio id municipio nombre  \
0    Egresos hospitalarios    2-2017-1293           NaN              NaN   
1  Ministerio de Seguridad       1-2017-1       60412.0      José C. Paz   
2  Ministerio de Seguridad    1-2017-1039       60412.0      José C. Paz   
3  Ministerio de Seguridad       1-2017-2       60412.0      José C. Paz   
4  Ministerio de Seguridad     1-2017-592       60098.0          Berisso   

  fecha hecho       hora hecho    momento hecho edad       sexo  \
0  14/12/2016  Sin especificar  Sin especificar   75  Masculino   
1  01/01/2017          7:10:00           Diurno   21  Masculino   
2  01/01/2017          6:05:00         Nocturno   22  Masculino   
3  01/01/2017          7:10:00           Diurno   42  Masculino   
4  01/01/2017          1:40:00         Nocturno   18  Masculino   

     clase victima vehiculo victima         tipo via  \
0  Sin especificar  Sin especificar  Sin especificar   
1        Conductor         V

In [8]:
# eliminar filas con valores nan
df_cleaned = df.dropna()

In [10]:
# Definir variables objetivo y features
X = df_cleaned.drop('clase victima', axis=1)  
y = df_cleaned['clase victima']


variable objetivo: numero de victimas
predecir cuantas victimas habra en un incidente especifico,basado en otras variables

In [21]:
# Convertir variables categóricas a variables dummy
X = pd.get_dummies(X, drop_first=True)


In [22]:
# divido los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [23]:
# modelo de clasificacion random forest
from sklearn.ensemble import RandomForestClassifier

# Inicializar el modelo
model_rf = RandomForestClassifier(random_state=42)

# Entrenar el modelo
model_rf.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

In [24]:
from sklearn.metrics import classification_report, accuracy_score

# Predicciones
y_pred_rf = model_rf.predict(X_test)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test, y_pred_rf))
print("Classification Report:\n", classification_report(y_test, y_pred_rf))


Accuracy: 0.7902097902097902
Classification Report:
                                       precision    recall  f1-score   support

                         Acompañante       0.55      0.26      0.35        23
                           Conductor       0.77      0.94      0.85        72
Ocupante de vehículo sin especificar       0.33      1.00      0.50         2
                            Pasajero       0.00      0.00      0.00         2
                              Peatón       1.00      0.82      0.90        22
                     Sin especificar       0.95      0.86      0.90        22

                            accuracy                           0.79       143
                           macro avg       0.60      0.65      0.58       143
                        weighted avg       0.78      0.79      0.77       143



C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


realizar la prediccion con xgboost

In [35]:
pip install xgboost scikit-learn pandas


Note: you may need to restart the kernel to use updated packages.


In [42]:
import pandas as pd

# Cargar el DataFrame especificando el delimitador como punto y coma
df = pd.read_csv('defuncionesviales.csv', delimiter=';', encoding='ISO-8859-1')

print(df.shape)  # Verifica que hay datos al cargar
print(df.head())  # Muestra las primeras filas del DataFrame


(1369, 14)
                    fuente numero victima  municipio id municipio nombre  \
0    Egresos hospitalarios    2-2017-1293           NaN              NaN   
1  Ministerio de Seguridad       1-2017-1       60412.0      José C. Paz   
2  Ministerio de Seguridad    1-2017-1039       60412.0      José C. Paz   
3  Ministerio de Seguridad       1-2017-2       60412.0      José C. Paz   
4  Ministerio de Seguridad     1-2017-592       60098.0          Berisso   

  fecha hecho       hora hecho    momento hecho edad       sexo  \
0  14/12/2016  Sin especificar  Sin especificar   75  Masculino   
1  01/01/2017          7:10:00           Diurno   21  Masculino   
2  01/01/2017          6:05:00         Nocturno   22  Masculino   
3  01/01/2017          7:10:00           Diurno   42  Masculino   
4  01/01/2017          1:40:00         Nocturno   18  Masculino   

     clase victima vehiculo victima         tipo via  \
0  Sin especificar  Sin especificar  Sin especificar   
1        Conducto

In [44]:
print(df.columns)


Index(['fuente', 'numero victima', 'municipio id', 'municipio nombre',
       'fecha hecho', 'hora hecho', 'momento hecho', 'edad', 'sexo',
       'clase victima', 'vehiculo victima', 'tipo via', 'tipo incidente',
       'clima'],
      dtype='object')


In [45]:
# Definir X e y
X = df.drop(['numero victima', 'clase victima'], axis=1)  # Eliminamos 'numero victima' y la variable objetivo
y = df['clase victima']  # Usamos 'clase victima' como variable objetivo


In [46]:
from sklearn.preprocessing import LabelEncoder

# Codificar la variable objetivo
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)


In [47]:
# Convertir columnas categóricas en variables dummy
X = pd.get_dummies(X, drop_first=True)


In [48]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import xgboost as xgb
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Cargar el DataFrame
df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='latin1')

# Definir X e y
X = df.drop(['numero victima', 'clase victima'], axis=1)  # Ajusta según las columnas
y = df['clase victima']  # Variable objetivo

# Codificar la variable objetivo
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

# Convertir columnas categóricas en variables dummy
X = pd.get_dummies(X, drop_first=True)

# Dividir el conjunto de datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Entrenar el modelo XGBoost
model = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model.fit(X_train, y_train)

# Evaluar el modelo
y_pred = model.predict(X_test)
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print('Classification Report:\n', classification_report(y_test, y_pred))
print('Confusion Matrix:\n', confusion_matrix(y_test, y_pred))
# Codificar la variable objetivo
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

C:\Users\metal\anaconda3\lib\site-packages\xgboost\core.py:158: UserWarning: [19:45:58] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0015a694724fa8361-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


Accuracy: 0.7956204379562044
Classification Report:
               precision    recall  f1-score   support

           0       0.72      0.49      0.58        53
           1       0.76      0.89      0.82       132
           2       1.00      1.00      1.00         3
           3       0.00      0.00      0.00         1
           4       1.00      1.00      1.00        22
           5       0.86      0.78      0.82        63

    accuracy                           0.80       274
   macro avg       0.72      0.69      0.70       274
weighted avg       0.79      0.80      0.79       274

Confusion Matrix:
 [[ 26  24   0   0   0   3]
 [  9 118   0   0   0   5]
 [  0   0   3   0   0   0]
 [  0   1   0   0   0   0]
 [  0   0   0   0  22   0]
 [  1  13   0   0   0  49]]


C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [71]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

# Cargar tus datos, especificando el delimitador y el encoding
try:
    df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='latin1')  # Intenta con 'latin1'
except UnicodeDecodeError:
    df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='cp1252')  # Intenta con 'cp1252'

# Mostrar las primeras filas y las columnas del DataFrame para verificar que se cargó correctamente
print(df.head())
print("Columnas disponibles:", df.columns.tolist())

# Definir la variable objetivo
variable_objetivo = 'clase víctima'  # Variable objetivo

if variable_objetivo in df.columns:
    # Definir las características (X) y la variable objetivo (y)
    X = df[['clima', 'edad', 'fuente', 'hora hecho']]  # Asegúrate de que estas columnas existan
    y = df[variable_objetivo]  # Usa 'clase víctima' como variable objetivo

    # Codificación de variables categóricas
    X = pd.get_dummies(X, drop_first=True)

    # Dividir el conjunto de datos en conjuntos de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Inicializar los modelos
    model_rf = RandomForestClassifier(random_state=42)
    model_lr = LogisticRegression(max_iter=1000)

    # Entrenar los modelos
    model_rf.fit(X_train, y_train)
    model_lr.fit(X_train, y_train)

    # Hacer predicciones
    y_pred_rf = model_rf.predict(X_test)  # Predicciones del modelo de bosque aleatorio
    y_pred_lr = model_lr.predict(X_test)  # Predicciones del modelo de regresión logística

    # Evaluar los modelos
    print("Random Forest Model:")
    print("Accuracy:", accuracy_score(y_test, y_pred_rf))
    print(classification_report(y_test, y_pred_rf))

    print("Logistic Regression Model:")
    print("Accuracy:", accuracy_score(y_test, y_pred_lr))
    print(classification_report(y_test, y_pred_lr))
else:
    print(f"La columna '{variable_objetivo}' no se encuentra en el DataFrame. Revisa las columnas disponibles.")


                    fuente numero victima  municipio id municipio nombre  \
0    Egresos hospitalarios    2-2017-1293           NaN              NaN   
1  Ministerio de Seguridad       1-2017-1       60412.0      José C. Paz   
2  Ministerio de Seguridad    1-2017-1039       60412.0      José C. Paz   
3  Ministerio de Seguridad       1-2017-2       60412.0      José C. Paz   
4  Ministerio de Seguridad     1-2017-592       60098.0          Berisso   

  fecha hecho       hora hecho    momento hecho edad       sexo  \
0  14/12/2016  Sin especificar  Sin especificar   75  Masculino   
1  01/01/2017          7:10:00           Diurno   21  Masculino   
2  01/01/2017          6:05:00         Nocturno   22  Masculino   
3  01/01/2017          7:10:00           Diurno   42  Masculino   
4  01/01/2017          1:40:00         Nocturno   18  Masculino   

     clase victima vehiculo victima         tipo via  \
0  Sin especificar  Sin especificar  Sin especificar   
1        Conductor         V

In [72]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

# Cargar tus datos, especificando el delimitador y el encoding
try:
    df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='latin1')  # Intenta con 'latin1'
except UnicodeDecodeError:
    df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='cp1252')  # Intenta con 'cp1252'

# Mostrar las primeras filas y las columnas del DataFrame para verificar que se cargó correctamente
print(df.head())
print("Columnas disponibles:", df.columns.tolist())

# Definir la variable objetivo
variable_objetivo = 'clase victima'  # Variable objetivo (ajustada)

if variable_objetivo in df.columns:
    # Definir las características (X) y la variable objetivo (y)
    X = df[['clima', 'edad', 'fuente', 'hora hecho']]  # Asegúrate de que estas columnas existan
    y = df[variable_objetivo]  # Usa 'clase victima' como variable objetivo

    # Codificación de variables categóricas
    X = pd.get_dummies(X, drop_first=True)

    # Dividir el conjunto de datos en conjuntos de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Inicializar los modelos
    model_rf = RandomForestClassifier(random_state=42)
    model_lr = LogisticRegression(max_iter=1000)

    # Entrenar los modelos
    model_rf.fit(X_train, y_train)
    model_lr.fit(X_train, y_train)

    # Hacer predicciones
    y_pred_rf = model_rf.predict(X_test)  # Predicciones del modelo de bosque aleatorio
    y_pred_lr = model_lr.predict(X_test)  # Predicciones del modelo de regresión logística

    # Evaluar los modelos
    print("Random Forest Model:")
    print("Accuracy:", accuracy_score(y_test, y_pred_rf))
    print(classification_report(y_test, y_pred_rf))

    print("Logistic Regression Model:")
    print("Accuracy:", accuracy_score(y_test, y_pred_lr))
    print(classification_report(y_test, y_pred_lr))
else:
    print(f"La columna '{variable_objetivo}' no se encuentra en el DataFrame. Revisa las columnas disponibles.")


                    fuente numero victima  municipio id municipio nombre  \
0    Egresos hospitalarios    2-2017-1293           NaN              NaN   
1  Ministerio de Seguridad       1-2017-1       60412.0      José C. Paz   
2  Ministerio de Seguridad    1-2017-1039       60412.0      José C. Paz   
3  Ministerio de Seguridad       1-2017-2       60412.0      José C. Paz   
4  Ministerio de Seguridad     1-2017-592       60098.0          Berisso   

  fecha hecho       hora hecho    momento hecho edad       sexo  \
0  14/12/2016  Sin especificar  Sin especificar   75  Masculino   
1  01/01/2017          7:10:00           Diurno   21  Masculino   
2  01/01/2017          6:05:00         Nocturno   22  Masculino   
3  01/01/2017          7:10:00           Diurno   42  Masculino   
4  01/01/2017          1:40:00         Nocturno   18  Masculino   

     clase victima vehiculo victima         tipo via  \
0  Sin especificar  Sin especificar  Sin especificar   
1        Conductor         V

C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metr

In [2]:
# Importar las librerías necesarias
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score
from sklearn.preprocessing import LabelEncoder
import numpy as np

# Cargar el archivo CSV con codificación 'latin1' y separador ';'
df = pd.read_csv('defuncionesviales.csv', sep=';', encoding='latin1')

# Verificar si hay valores NaN
print("Valores NaN antes de la imputación:\n", df.isnull().sum())

# Opciones para manejar valores NaN: 
# Puedes imputarlos con la media, moda, o eliminarlos.
# En este ejemplo, se imputarán con la media (para valores numéricos) o la moda (para valores categóricos).

# Imputar valores NaN numéricos con la media y categóricos con la moda
for column in df.columns:
    if df[column].dtype == np.number:
        df[column].fillna(df[column].mean(), inplace=True)
    else:
        df[column].fillna(df[column].mode()[0], inplace=True)

# Verificar que no haya más valores NaN
print("Valores NaN después de la imputación:\n", df.isnull().sum())

# Codificar las variables categóricas con LabelEncoder (si es necesario)
le = LabelEncoder()

# Aplicar LabelEncoder a todas las columnas categóricas
for col in df.select_dtypes(include=['object']).columns:
    df[col] = le.fit_transform(df[col])

# Separar las características (X) y la variable objetivo (y)
X = df.drop('clase victima', axis=1)  # Características
y = df['clase victima']  # Variable objetivo

# Dividir el dataset en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Inicializar los modelos
modelos = {
    "Random Forest": RandomForestClassifier(),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='mlogloss'),
    "Logistic Regression": LogisticRegression(max_iter=1000)
}

# Entrenar y evaluar cada modelo
resultados = {}
for nombre, modelo in modelos.items():
    print(f"\nEntrenando y evaluando el modelo: {nombre}")
    modelo.fit(X_train, y_train)
    y_pred = modelo.predict(X_test)
    
    # Evaluar las métricas de cada modelo
    accuracy = accuracy_score(y_test, y_pred)
    report = classification_report(y_test, y_pred)
    
    resultados[nombre] = {
        'accuracy': accuracy,
        'report': report
    }

# Imprimir los resultados comparativos
for nombre, resultado in resultados.items():
    print(f"\nModelo: {nombre}")
    print(f"Exactitud (Accuracy): {resultado['accuracy']}")
    print("Reporte de Clasificación:")
    print(resultado['report'])


Valores NaN antes de la imputación:
 fuente                0
numero victima        0
municipio id        656
municipio nombre    656
fecha hecho           0
hora hecho            0
momento hecho         0
edad                  0
sexo                  0
clase victima         0
vehiculo victima      0
tipo via              0
tipo incidente        0
clima                 0
dtype: int64
Valores NaN después de la imputación:
 fuente              0
numero victima      0
municipio id        0
municipio nombre    0
fecha hecho         0
hora hecho          0
momento hecho       0
edad                0
sexo                0
clase victima       0
vehiculo victima    0
tipo via            0
tipo incidente      0
clima               0
dtype: int64

Entrenando y evaluando el modelo: Random Forest

Entrenando y evaluando el modelo: XGBoost


C:\Users\metal\AppData\Local\Temp\ipykernel_20548\2754328950.py:23: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if df[column].dtype == np.number:
C:\Users\metal\AppData\Local\Temp\ipykernel_20548\2754328950.py:23: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if df[column].dtype == np.number:
C:\Users\metal\AppData\Local\Temp\ipykernel_20548\2754328950.py:23: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if df[column].dtype == np.number:
C:\Users\metal\AppData\Local\Temp\ipykernel_20548\2754328950.py:23: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if df[column].dtyp


Entrenando y evaluando el modelo: Logistic Regression

Modelo: Random Forest
Exactitud (Accuracy): 0.8004866180048662
Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.66      0.47      0.55        74
           1       0.76      0.90      0.82       201
           2       1.00      1.00      1.00         7
           3       0.00      0.00      0.00         2
           4       0.97      1.00      0.99        37
           5       0.91      0.78      0.84        90

    accuracy                           0.80       411
   macro avg       0.72      0.69      0.70       411
weighted avg       0.80      0.80      0.79       411


Modelo: XGBoost
Exactitud (Accuracy): 0.8126520681265207
Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.64      0.55      0.59        74
           1       0.80      0.87      0.83       201
           2       1.00      1.00      1.00         7
           

C:\Users\metal\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\metal\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in l

##Conclusión:
XGBoost tiene el mejor rendimiento en términos de exactitud y generalización (f1-score ponderado de 0.81), seguido de cerca por Random Forest.
Logistic Regression muestra un rendimiento significativamente peor que los otros dos modelos, con una precisión y recuperación bajas en la mayoría de las clases.
Recomendación: Dado que XGBoost ofrece una ligera mejora en el rendimiento, podrías considerarlo como el mejor modelo para este conjunto de datos, aunque Random Forest también es competitivo.